src/NuGet.Core/NuGet.Protocol.Core.v3/PackageUpdateResource.cs

```diff
@@ -0,0 +1,601 @@
+
+        private static IEnumerable<string> GetPackagesToPush(string packagePath)
+        {
+            // Ensure packagePath ends with *.nupkg
+            packagePath = EnsurePackageExtension(packagePath);
+            return PerformWildcardSearch(Directory.GetCurrentDirectory(), packagePath);
+        }
+
+        private static string EnsurePackageExtension(string packagePath)
+        {
+            if (packagePath.IndexOf('*') == -1)
+            {
+                // If there's no wildcard in the path to begin with, assume that it's an absolute path.
+                return packagePath;
+            }
+            // If the path does not contain wildcards, we need to add *.nupkg to it.
+            if (!packagePath.EndsWith(NuGetConstants.PackageExtension, StringComparison.OrdinalIgnoreCase))
+            {
+                if (packagePath.EndsWith("**", StringComparison.OrdinalIgnoreCase))
+                {
+                    packagePath = packagePath + Path.DirectorySeparatorChar + '*';
+                }
+                else if (!packagePath.EndsWith("*", StringComparison.OrdinalIgnoreCase))
+                {
+                    packagePath = packagePath + '*';
+                }
+                packagePath = packagePath + NuGetConstants.PackageExtension;
+            }
+            return packagePath;
+        }
+
+        private static void EnsurePackageFileExists(string packagePath, IEnumerable<string> packagesToPush)
+        {
+            if (!packagesToPush.Any())
+            {
+                throw new ArgumentException(String.Format(CultureInfo.CurrentCulture,
+                    Strings.UnableToFindFile,
+                    packagePath));
+            }
+        }
+
+        // Indicates whether the specified source is a file source, such as: \\a\b, c:\temp, etc.
+        private static bool IsFileSource(string source)
+        {
+            Uri uri;
+            if (Uri.TryCreate(source, UriKind.RelativeOrAbsolute, out uri))
+            {
+                return uri.IsFile;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        // Pushes a package to the Http server.
+        private async Task PushPackageToServer(string source,
+            string apiKey,
+            string pathToPackage,
+            long packageSzie,
+            ILogger logger,
+            CancellationToken token)
+        {
+            HttpRequestMessage request = null;
+            HttpResponseMessage response = null;
+
+            try
+            {
+                response = await _httpSource.SendAsync(
+                    () => { return request = CreateRequest(request, source, pathToPackage, apiKey); },
+                    token);
+            }
+            finally
+            {
+                if (request != null)
+                {
+                    request.Dispose();
+                }
+            };
+
+            response.EnsureSuccessStatusCode();
```

Comment:
This can be done inside a using around the reponse. The whole try/finally will not be needed.